File: minimum-number-of-moves-to-make-palindrome/solution2193.py
# https://leetcode.com/problems/minimum-number-of-moves-to-make-palindrome/

class Solution2193:

    # T: O(n), S: O(n)
    def minMovesToMakePalindrome(self, s: str) -> int:
        moves = 0
        l, r = 0, len(s) - 1
        arr = list(s)
        while l < r:
            m = r
            while m > l and arr[m] != arr[l]:
                m -= 1
            if m == l:
                arr[l], arr[l + 1] = arr[l + 1], arr[l]
                moves += 1
                continue
            while m < r:
                arr[m], arr[m + 1] = arr[m + 1], arr[m]
                m += 1
                moves += 1
            l += 1
            r -= 1

        return moves


def main() -> None:
    sol = Solution2193()
    s1 = "aabb"
    s2 = "letelt"
    print(sol.minMovesToMakePalindrome(s1))  # 2
    print(sol.minMovesToMakePalindrome(s2))  # 2


if __name__ == "__main__":
    main()

File: minimum-number-of-moves-to-make-palindrome/Solution2193.java
// https://leetcode.com/problems/minimum-number-of-moves-to-make-palindrome/

class Solution2193 {

    // T: O(n), S: O(n)
    public int minMovesToMakePalindrome(String s) {
        int moves = 0, l = 0, r = s.length() - 1;
        var arr = s.toCharArray();
        while (l < r) {
            int m = r;
            // Find the matching character for arr[l] from the end
            while (m > l && arr[m] != arr[l]) {
                m--;
            }
            // Odd length string's mid character, move to middle
            if (m == l) {
                swap(arr, l, l + 1);
                moves++;
                continue;
            }
            // Move the matching character to the correct position
            while (m < r) {
                swap(arr, m, m + 1);
                m++;
                moves++;
            }
            l++;
            r--;
        }
        return moves;
    }

    private static void swap(char[] arr, int a, int b) {
        char temp = arr[a];
        arr[a] = arr[b];
        arr[b] = temp;
    }

    public static void main(String[] args) {
        var sol = new Solution2193();
        var s1 = "aabb";
        var s2 = "letelt";
        System.out.println(sol.minMovesToMakePalindrome(s1)); // 2
        System.out.println(sol.minMovesToMakePalindrome(s2)); // 2
    }

}
